aten/src/ATen/native/TensorIndexing.h

```diff
@@ -108,5 +111,170 @@ struct CAFFE2_API TensorIndex final {
 CAFFE2_API std::ostream& operator<<(std::ostream& stream, const TensorIndex& tensor_index);
 CAFFE2_API std::ostream& operator<<(std::ostream& stream, const std::vector<TensorIndex>& tensor_indices);
 
+inline Tensor applySlice(
+    const Tensor& self,
+    int64_t dim,
+    int64_t start,
+    int64_t stop,
+    int64_t step,
+    bool ensure_view=false,
+    bool is_tracing=false) {
+  const auto& length = self.size(dim);
+
+  TORCH_CHECK_VALUE(step != 0, "step cannot be zero");
+  // TODO: implement negative step
+  TORCH_CHECK_VALUE(step >= 0, "negative step not yet supported");
+
+  // Skip this optimization if we are tracing, as the trace may be polymorphic
+  // over the shape of the `self` tensor, and we still want to record
+  // the slice.
+  if (!ensure_view && start == 0 && stop == length && step == 1 && !is_tracing) {
+    return self;
+  }
+  return self.slice(dim, start, stop, step);
+}
+
+inline Tensor applySelect(const Tensor& self, int64_t dim, int64_t index, int64_t real_dim=0) {
+  TORCH_CHECK_INDEX(
+    !(index == 0 && dim == 0 && self.dim() == 0),
+    "invalid index of a 0-dim tensor. ",
+    "Use tensor.item() to convert a 0-dim tensor to a number");
+
+  int64_t size = self.size(dim);
+  TORCH_CHECK_INDEX(
+    index >= -size && index < size,
+    "index ", index, " is out of bounds for dimension ", real_dim, " with size ", size);
+
+  // if the index is negative, do not normalize it because that would fix the index
+  // on the current tensor size in the tracer.
+  // aten::select also works on negative indices
+  return self.select(dim, index);
+}
+
+inline Tensor boolToIndexingTensor(const Tensor& self, bool value) {
+  // booleans add a dimension of size 1. true indexes this dimension as if 0:, false as empty.
+  if (value) {
+    return at::native::zeros({1}, {}, self.options().dtype(kLong));
+  } else {
+    return at::native::empty({0}, {}, self.options().dtype(kLong));
+  }
+}
+
+inline void recordTensorIndex(const Tensor& tensor, std::vector<Tensor>& outIndices, int64_t& dim) {
+  // TODO: check scalarType
+  outIndices.resize(dim + 1);
+  outIndices[dim] = tensor;
+  dim++;
+};
+
+inline Tensor handleInteger(const Tensor& self, int64_t& dim, int64_t index, int64_t real_dim) {
+  return applySelect(self, dim, index, real_dim);
+}
+
+inline Tensor handleSlice(const Tensor& self, int64_t& dim, int64_t start, int64_t stop, int64_t step, bool is_tracing=false) {
```

Comment:
The mutable reference embedded in the middle of the argument list here is really poor C++ style. In general you should avoid mutable reference arguments (for example, Google C++ forbids them entirely--PyTorch codebase has a few, so it's not quite a blanket denial). In the contexts where handleSlice is called, it would be much, much clearer to take an `int64_t*`